SeQuant/core/export/python_einsum.hpp

```diff
@@ -0,0 +1,636 @@
+    }
+    // Default fallback - use a simple naming scheme
+    return "dim_" + toUtf8(space.base_key());
+  }
+
+  /// Get the shape tuple for a tensor
+  std::string get_shape_tuple(const Tensor &tensor) const {
+    std::string shape = "(";
+    bool first = true;
+    for (const Index &idx : tensor.const_indices()) {
+      if (!first) shape += ", ";
+      shape += get_shape(idx.space());
+      first = false;
+    }
+    shape += ")";
+    return shape;
+  }
+
+  /// Set the dimension/shape for a given index space
+  void set_shape(const IndexSpace &space, std::string shape) {
+    m_index_shapes[space] = std::move(shape);
+  }
+
+  /// Get the backend being used
+  PythonEinsumBackend backend() const { return m_backend; }
+
+  /// Set the backend
+  void set_backend(PythonEinsumBackend backend) { m_backend = backend; }
+
+  /// Get the module prefix (np. or torch.)
+  std::string module_prefix() const {
+    return m_backend == PythonEinsumBackend::NumPy ? "np." : "torch.";
+  }
+
+ protected:
+  ShapeMap m_index_shapes;
+  PythonEinsumBackend m_backend;
+};
+
+/// Generator for producing Python code using numpy.einsum or torch.einsum
+/// - NumPy: https://numpy.org/doc/stable/reference/generated/numpy.einsum.html
+/// - PyTorch: https://pytorch.org/docs/stable/generated/torch.einsum.html
+template <typename Context = PythonEinsumGeneratorContext>
+class PythonEinsumGenerator : public Generator<Context> {
+ public:
+  PythonEinsumGenerator() = default;
+  ~PythonEinsumGenerator() = default;
+
+  std::string get_format_name() const override { return "Python (einsum)"; }
+
+  bool supports_named_sections() const override { return true; }
+
+  bool requires_named_sections() const override { return false; }
+
+  DeclarationScope index_declaration_scope() const override {
+    return DeclarationScope::Global;
+  }
+
+  DeclarationScope variable_declaration_scope() const override {
+    return DeclarationScope::Global;
+  }
+
+  DeclarationScope tensor_declaration_scope() const override {
+    return DeclarationScope::Global;
+  }
+
+  PrunableScalars prunable_scalars() const override {
+    return PrunableScalars::All;
+  }
+
+  std::string represent(const Index &idx, const Context &ctx) const override {
+    if (idx.has_proto_indices()) {
+      throw std::runtime_error("Proto Indices are not (yet) supported!");
+    }
+
+    // Convert index label to a single character for einsum notation
+    // For einsum, we need single characters
+    std::string label = toUtf8(idx.full_label());
+
+    // Try to extract a single character representation
+    // If the label is longer, use its first character (this may need
+    // customization)
+    if (label.empty()) {
+      throw std::runtime_error("Empty index label");
+    }
+
+    // Return first character (lowercase for einsum convention)
+    char c = label[0];
+    if (std::isupper(static_cast<unsigned char>(c))) {
+      c = std::tolower(static_cast<unsigned char>(c));
+    }
```

Comment:
```suggestion
    c = std::tolower(static_cast<unsigned char>(c));
```
I believe, it is always save to call `tolower`